src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -128,10 +131,17 @@ public function authenticate(RequestEvent $event)
         }
 
         if ($token instanceof TokenInterface) {
+            $originalToken = $token;
             $token = $this->refreshUser($token);
 
-            if (!$token && $this->rememberMeServices) {
-                $this->rememberMeServices->loginFail($request);
+            if (!$token) {
+                if ($this->dispatcher) {
+                    $this->dispatcher->dispatch(new TokenDeauthenticatedEvent($originalToken, $request));
+                }
+
+                if ($this->rememberMeServices) {
+                    $this->rememberMeServices->loginFail($request);
+                }
```

Comment:
Would passing `RememberMeServices` to this class be deprecated later?

Reply:
Yes, but we can only do so when deprecating the legacy system (as this is completely valid in the legacy system).